crates/conformance-test-utils/src/madara.rs

```diff
@@ -0,0 +1,154 @@
+use eyre::Result;
+use kakarot_rpc_core::client::constants::STARKNET_NATIVE_TOKEN;
+use reth_primitives::{U128, U256};
+use starknet::core::types::FieldElement;
+use starknet::core::utils::get_storage_var_address;
+
+use crate::types::{ContractAddress, StorageKey, StorageValue};
+
+/// Generates the genesis storage tuple for setting the storage of a Starknet contract.
+///
+/// This function calculates the storage key for the storage variable `storage_variable_name` and
+/// its keys. The resulting tuple represents the initial storage of the contract, where the storage
+/// key at a given `storage_key_offset` is set to the specified `storage_value`.
+pub fn genesis_set_storage_starknet_contract(
+    starknet_address: FieldElement,
+    storage_variable_name: &str,
+    keys: &[FieldElement],
+    storage_value: FieldElement,
+    storage_key_offset: u64,
+) -> Result<((ContractAddress, StorageKey), StorageValue)> {
+    // Compute the storage key for the storage variable name and the keys.
+    let mut storage_key = get_storage_var_address(storage_variable_name, keys)?;
+
+    // Add the offset to the storage key.
+    storage_key += FieldElement::from(storage_key_offset);
+
+    let contract_address: ContractAddress = starknet_address.into();
+
+    // Create the tuple for the initial storage data on the Kakarot contract with the given storage key.
+    let storage_data = ((contract_address, storage_key.into()), storage_value.into());
+
+    Ok(storage_data)
+}
+
+/// Generates the genesis storage tuples for pre-funding a Starknet address on Starknet.
+///
+/// This function calculates the storage keys for the balance of the ERC20 Fee Token
+/// contract using the provided Starknet address. The resulting Vec of tuples represent the initial
+/// storage of the Fee Token contract, where the account associated with the Starknet address is
+/// pre-funded with the specified `amount`. The `amount` is split into two 128-bit chunks, which
+/// are stored in the storage keys at offsets 0 and 1.
+pub fn genesis_fund_starknet_address(
+    starknet_address: FieldElement,
+    amount: U256,
+) -> Result<Vec<((ContractAddress, StorageKey), StorageValue)>> {
+    // Split the amount into two 128-bit chunks.
+    let low = amount & U256::from(U128::MAX);
+    let high = amount >> 128;
+
+    // The storage key offsets for the two 128-bit chunks.
+    let uint256_offset = [(low, 0), (high, 1)]; // (value, offset)
+
+    // Iterate over the storage key offsets and generate the storage tuples.
+    uint256_offset
+        .iter()
+        .filter(|(value, _)| *value != U256::from(0)) // Only add non-zero balances.
```

Comment:
why do we need this check? what's wrong with 0 balances?

Reply:
I was thinking by default the storage value is 0 so we don't need to generate the storage data for it, but for completeness I can remove this 👍